README.md
# Noto Emoji SVG

OpenType-SVG version of [Noto Emoji](https://github.com/googlefonts/noto-emoji). OpenType-SVG is a font format that has some or all glyphs represented as SVG (scalable vector graphics) artwork. This allows glyphs to be displayed in multiple colors and gradients. OpenType-SVG fonts are also known as "color fonts". For more information, visit https://color.typekit.com. 


## License

As a derivative of [Noto Emoji](https://github.com/googlefonts/noto-emoji), the same licenses apply for Noto Emoji SVG:  
Emoji fonts (in the fonts/ subdirectory) are under the
[SIL Open Font License, version 1.1](fonts/LICENSE).  
Tools and most image resources are under the [Apache license, version 2.0](./LICENSE).
Flag images under flags/ are in the public domain or
otherwise exempt from copyright ([more info](https://github.com/googlefonts/noto-emoji/blob/main/third_party/region-flags/LICENSE)).


### ⚠️ NOTE ⚠️

The repository contains aliases/symlinks. Cloning on Windows
requires using the option `-c core.symlinks=true` and running the command with
Admin privileges.

	git -c core.symlinks=true clone https://github.com/adobe-fonts/noto-emoji-svg.git

You may also need to pull using this command,

	git -c core.symlinks=true pull

and configure the repository with this command,

	git config core.symlinks true


### ⚠️ NOTE 2 ⚠️

As of July 9th 2022, browsers that do not support OT-SVG fonts
are **Google Chrome** (tested in v103.0) and **Opera** (tested in v89.0).

**Firefox** supports OT-SVG (tested in v102.0). A former bug where non-1000 UPM fonts were displayed at the wrong size has been fixed.

**Safari** also supports OT-SVG, but the number of SVG elements on a page is limited (tested in v15.5). SVGs on the test.html page will drop out after approximately 800 test samples (which implies 3200 SVGs rendered). This is why the test has been split into multiple html files.


### Requirements

* Python 3.6+
* FontTools
* AFDKO


## Building the fonts

When building the **black-and-white font** and **color font**, make sure to build the **black-and-white font** first before building the **color font**. Since the data in the **color font** is dependent on the data in the **black-and-white font**, the **color font** will not build properly if it is built first. 

To build the **black-and-white font** use the following command,

	python3 make_bw_font.py -o fonts -r x.xxx --gsub GSUB.fea --gpos GPOS.fea --uvs UVS.txt svg_bw flags_bw

and to build the **color font** use the following command,

	python3 make_svg_font.py -z -r x.xxx svg flags fonts/NotoEmoji.otf -v

where `x.xxx` is the version number to be assigned to the font (e.g `1.082`).

It's possible to build both fonts (and to subroutinize the BW font) using the shell
script that combines all commands,

	sh build.sh x.xxx


## Subroutinizing the OTFs

The OT-CFF fonts can be subroutinized with the following command:

	sh subroutinize.sh <font file path>

Subroutinizing requires AFDKO's `tx` and `sfntedit` tools.


## Generating the HTML test document

To help thoroughly test the fonts, a script was developed that generates
[an HTML file](test.html) that lists all official emoji and emoji sequences.
The current output shows Unicode version 13.1. To generate the file run this command:

	python3 test/generate_test_html.py

To divide the output into multiple files use the `--paginate` option:

	python3 test/generate_test_html.py -p

To generate a subset of the **test.html** file —named **test-changes.html**— containing only the changes from the previous font build, use the `-c/--changes` option. Keep in mind that this option depends on the contents of the [changes.txt](test/changes.txt) file.


## Generating the TXT test document

Another script was developed that generates the test document in simple TXT format.
To generate the [TXT test file](test.txt) run this command:

	python3 test/generate_test_txt.py -e 20

This file can be easily imported into Illustrator, Photoshop or InDesign.

The [emoji-test.txt](test/emoji-test.txt) file was obtained from http://unicode.org/Public/emoji/

As with the HTML test document, it's possible to generate a subset file named
**test-changes.txt** by using the `-c/--changes` option.


## Generating aliases

Aliases/symbolic links are used extensively to avoid having multiple copies of the same artwork.

To generate **SVG black-and-white aliases** run this command:

	python3 make_aliases.py emoji_bw_aliases.txt svg_bw

To generate **SVG color aliases** run this command:

	python3 make_aliases.py emoji_color_aliases.txt svg

To generate **PNG color aliases** run this command:

	python3 make_aliases.py emoji_color_aliases.txt png

To generate **SVG black-and-white flag aliases** run this command:

	python3 make_aliases.py flag_bw_aliases.txt flags_bw

To generate **SVG color flag aliases** run this command:

	python3 make_aliases.py flag_color_aliases.txt flags

To generate **PNG color flag aliases** run this command:

	python3 make_aliases.py flag_color_aliases.txt flags_png


## Generating PNG and SVG files

In the event that only the original Ai artwork files are provided, use the
following steps to generate PNG or SVG files:

1. Open the Adobe Illustrator file(s)
2. Select the menu File > Scripts > Other Script...
3. Pick one of the scripts from the folder [ai_scripts](ai_scripts)
4. Select a folder to save the output files into


## Cleaning/sanitizing the SVG artwork

Whenever the SVG files are edited and saved with Adobe Illustrator they contain
unnecessary data that can and should be removed.

To clean the **black-and-white artwork** run this command:

	python3 svg_cleaner.py -k bw svg_bw

To clean the **color artwork** run this command:

	python3 svg_cleaner.py svg

To remove any CSS in-line styles from all SVG files (CSS is incompatible with macOS in this context), run:

	python3 svg_css_replacement.py


## Adobe Illustrator saving options

![SVG save options](Ai_save_options.png)


test/changes.txt
############################################
# NOTE: Do NOT put entries tagged as       #
#       - 'minimally-qualified',           #
#       - 'non-fully-qualified' or         #
#       - 'unqualified'                    #
#       in here, since they will be        #
#       ignored by the processing scripts. #
############################################


1F9CC                                                  ; fully-qualified     # 🧌 E14.0 troll
1FAB7                                                  ; fully-qualified     # 🪷 E14.0 lotus
1FAB8                                                  ; fully-qualified     # 🪸 E14.0 coral
1FAB9                                                  ; fully-qualified     # 🪹 E14.0 empty nest
1FABA                                                  ; fully-qualified     # 🪺 E14.0 nest with eggs
1FAD7                                                  ; fully-qualified     # 🫗 E14.0 pouring liquid
1FAD9                                                  ; fully-qualified     # 🫙 E14.0 jar
1F6DE                                                  ; fully-qualified     # 🛞 E14.0 wheel
1F6DD                                                  ; fully-qualified     # 🛝 E14.0 playground slide
1F6DF                                                  ; fully-qualified     # 🛟 E14.0 ring buoy
1FAAC                                                  ; fully-qualified     # 🪬 E14.0 hamsa
1FAA9                                                  ; fully-qualified     # 🪩 E14.0 mirror ball
1FA7C                                                  ; fully-qualified     # 🩼 E14.0 crutch
1FA7B                                                  ; fully-qualified     # 🩻 E14.0 x-ray
1F7F0                                                  ; fully-qualified     # 🟰 E14.0 heavy equals sign
1FAE0                                                  ; fully-qualified     # 🫠 E14.0 melting face
1FAE2                                                  ; fully-qualified     # 🫢 E14.0 face with open eyes and hand over mouth
1FAE3                                                  ; fully-qualified     # 🫣 E14.0 face with peeking eye
1FAE1                                                  ; fully-qualified     # 🫡 E14.0 saluting face
1FAE5                                                  ; fully-qualified     # 🫥 E14.0 dotted line face
1FAE4                                                  ; fully-qualified     # 🫤 E14.0 face with diagonal mouth
2639                                                   ; unqualified         # ☹ E0.7 frowning face
1F626                                                  ; fully-qualified     # 🥹 E14.0 face holding back tears
1FAE6                                                  ; fully-qualified     # 🫦 E14.0 biting lip

1FAC5                                                  ; fully-qualified     # 🫅 E14.0 person with crown
1FAC5 1F3FB                                            ; fully-qualified     # 🫅🏻 E14.0 person with crown: light skin tone
1FAC5 1F3FC                                            ; fully-qualified     # 🫅🏼 E14.0 person with crowne: medium-light skin tone
1FAC5 1F3FD                                            ; fully-qualified     # 🫅🏽 E14.0 person with crown: medium skin tone
1FAC5 1F3FE                                            ; fully-qualified     # 🫅🏾 E14.0 person with crown: medium-dark skin tone
1FAC5 1F3FF                                            ; fully-qualified     # 🫅🏿 E14.0 person with crown: dark skin tone
1FAC3                                                  ; fully-qualified     # 🫃 E14.0 pregnant man
1FAC3 1F3FB                                            ; fully-qualified     # 🫃🏻 E14.0 pregnant man: light skin tone
1FAC3 1F3FC                                            ; fully-qualified     # 🫃🏼 E14.0 pregnant man: medium-light skin tone
1FAC3 1F3FD                                            ; fully-qualified     # 🫃🏽 E14.0 pregnant man: medium skin tone
1FAC3 1F3FE                                            ; fully-qualified     # 🫃🏾 E14.0 pregnant man: medium-dark skin tone
1FAC3 1F3FF                                            ; fully-qualified     # 🫃🏿 E14.0 pregnant man: dark skin tone
1FAC4                                                  ; fully-qualified     # 🫄 E14.0 pregnant person
1FAC4 1F3FB                                            ; fully-qualified     # 🫄🏻 E14.0 pregnant person: light skin tone
1FAC4 1F3FC                                            ; fully-qualified     # 🫄🏼 E14.0 pregnant person: medium-light skin tone
1FAC4 1F3FD                                            ; fully-qualified     # 🫄🏽 E14.0 pregnant person: medium skin tone
1FAC4 1F3FE                                            ; fully-qualified     # 🫄🏾 E14.0 pregnant person: medium-dark skin tone
1FAC4 1F3FF                                            ; fully-qualified     # 🫄🏿 E14.0 pregnant person: dark skin tone
1FAF1                                                  ; fully-qualified     # 🫱 E14.0 rightwards hand
1FAF1 1F3FB                                            ; fully-qualified     # 🫱🏻 E14.0 rightwards hand: light skin tone
1FAF1 1F3FC                                            ; fully-qualified     # 🫱🏼 E14.0 rightwards hand: medium-light skin tone
1FAF1 1F3FD                                            ; fully-qualified     # 🫱🏽 E14.0 rightwards hand: medium skin tone
1FAF1 1F3FE                                            ; fully-qualified     # 🫱🏾 E14.0 rightwards hand: medium-dark skin tone
1FAF1 1F3FF                                            ; fully-qualified     # 🫱🏿 E14.0 rightwards hand: dark skin tone
1FAF2                                                  ; fully-qualified     # 🫲 E14.0 leftwards hand
1FAF2 1F3FB                                            ; fully-qualified     # 🫲🏻 E14.0 leftwards hand: light skin tone

test/generate_test_txt.py
# Copyright © 2019 Adobe, Inc.
# Author: Miguel Sousa
"""
Generates a simple text file for testing emoji characters.
"""
import argparse
import os
import sys

from generate_test_html import (
    append_to_file,
    parse_emoji_test_file,
    CHANGES_INPUT_PATH,
    TEST_DIR,
    TEST_INPUT_PATH,
)

TEST_OUTPUT_FILENAME = 'test.txt'
CHANGES_OUTPUT_FILENAME = 'test-changes.txt'


def positive_int(int_str):
    try:
        num_items = int(int_str)
    except ValueError:
        raise argparse.ArgumentTypeError(
            "'{}' is not an integer.".format(int_str))
    if num_items < 1:
        raise argparse.ArgumentTypeError('Number of emoji per line must be 1 '
                                         'or more.')
    return num_items


def main(args=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        '-e',
        '--emoji-per-line',
        help=('split the single line of emoji characters/sequences by '
              'introducing line breaks'),
        metavar='INTEGER',
        type=positive_int,
    )
    parser.add_argument(
        '-s',
        '--space',
        help=('add space character between emoji characters/sequences'),
        action='store_true'
    )
    parser.add_argument(
        '-c',
        '--changes',
        help="generates '{}' instead of '{}'".format(
            CHANGES_OUTPUT_FILENAME, TEST_OUTPUT_FILENAME),
        action='store_true',
    )
    opts = parser.parse_args(args)

    if opts.changes:
        emoji_input_path = CHANGES_INPUT_PATH
        emoji_output_filename = CHANGES_OUTPUT_FILENAME
    else:
        emoji_input_path = TEST_INPUT_PATH
        emoji_output_filename = TEST_OUTPUT_FILENAME

    # collect the list of codepoints
    cdpts_list = parse_emoji_test_file(emoji_input_path)

    # start a new file (avoids appending to an existing file)
    test_file_path = os.path.join(TEST_DIR, '..', emoji_output_filename)
    open(test_file_path, 'w').close()

    # begin the file with the BOM and use utf-16-le encoding
    # (to make Adobe Illustrator happy)
    append_to_file(test_file_path, '\uFEFF', 'utf-16-le')

    emoji_list = []
    for i, cps in enumerate(cdpts_list, 1):
        emoji = ''.join(chr(int(cp, 16)) for cp in cps)
        emoji_list.append(emoji)

    if opts.space:
        space = ' '
    else:
        space = ''

    if opts.emoji_per_line:
        emoji_stream = ''
        for i, emoji_group in enumerate(emoji_list, 1):
            emoji_sep = '\r' if i % opts.emoji_per_line == 0 else space
            emoji_stream += '{}{}'.format(emoji_group, emoji_sep)
    else:
        emoji_stream = space.join(emoji_list)

    append_to_file(test_file_path, emoji_stream, 'utf-16-le')


if __name__ == "__main__":
    sys.exit(main())


make_aliases.py
# Copyright © 2019 Adobe, Inc.
# Author: Miguel Sousa
"""
Creates aliases of SVG or PNG files in the same directory.
"""
import argparse
import io
import logging
import os
import sys

from make_bw_font import validate_dir_path, validate_file_path


FILE_PREFIX = 'u'
FILE_EXTENSIONS = ('svg', 'png')

log = logging.getLogger('make_aliases')


def sniff_file_extension(src_name):
    for ext in FILE_EXTENSIONS:
        src_filename = '{}{}.{}'.format(FILE_PREFIX, src_name, ext)
        if os.path.exists(src_filename):
            return src_filename, ext
    # no file was found
    return None, None


def make_aliases(aliases_list, in_dir):
    os.chdir(in_dir)

    for src_name, dst_name in aliases_list:
        src_filename, ext = sniff_file_extension(src_name)
        if not src_filename:
            log.warning("File named '{}{}' not found in '{}'".format(
                FILE_PREFIX, src_name, in_dir))
            continue

        dst_filename = '{}{}.{}'.format(FILE_PREFIX, dst_name, ext)
        if os.path.exists(dst_filename):
            os.remove(dst_filename)

        try:
            os.symlink(src_filename, dst_filename)
        except OSError as err:
            if err.args == ('symbolic link privilege not held',):
                log.error('On Windows this script must be run in Admin mode.')
            else:
                print(src_filename)
                log.error('Failure while trying to create alias/symlink.')
            return 1


def parse_aliases_file(file_path):
    """
    Parses an emoji aliases text file.
    Returns a list of tuples in the form ('src_name', 'dst_name').
    """
    with io.open(file_path, encoding='utf-8') as fp:
        lines = fp.read().splitlines()

    aliases_list = []
    for line in lines:
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        # strip in-line comments
        comment_idx = line.find('#')
        if comment_idx > 0:
            line = line[:comment_idx].strip()
        aliases_list.append(tuple(line.split(';')))
    return aliases_list


def main(args=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        '-v',
        '--verbose',
        help='verbose mode. Use -vv for debug mode',
        action='count',
        default=0
    )
    parser.add_argument(
        'in_file',
        help='text file containing a listing of the aliases',
        metavar='FILE',
        type=validate_file_path,
    )
    parser.add_argument(
        'in_dir',
        help='input directory containing SVG files',
        metavar='DIR',
        type=validate_dir_path,
    )
    opts = parser.parse_args(args)

    if not opts.verbose:
        level = "WARNING"
    elif opts.verbose == 1:
        level = "INFO"
    else:
        level = "DEBUG"
    logging.basicConfig(level=level)

    aliases_list = parse_aliases_file(opts.in_file)
    make_aliases(aliases_list, opts.in_dir)


if __name__ == "__main__":
    sys.exit(main())


make_bw_font.py
# Copyright © 2019 Adobe, Inc.
# Author: Miguel Sousa
"""
Creates a sans-color emoji OT-CFF font from b&w SVG files.
"""
import argparse
from ast import literal_eval
from collections import deque
import glob
import io
import logging
import os
import re
import sys

from fontTools.feaLib.builder import addOpenTypeFeatures
from fontTools.fontBuilder import FontBuilder
from fontTools.misc.psCharStrings import T2CharString
from fontTools.pens.t2CharStringPen import T2CharStringPen
from fontTools.svgLib.path import SVGPath

COPYRIGHT = 'Copyright 2013 Google Inc.'
TRADEMARK = 'Noto is a trademark of Google Inc.'
FAMILY_NAME = 'Noto Emoji'
STYLE_NAME = 'Regular'
FULL_NAME = FAMILY_NAME
PS_NAME = 'NotoEmoji'
MANUFACTURER = 'Google Inc. & Adobe Inc.'
DESIGNER = 'Google Inc.'
VENDOR = 'GOOG'
VENDOR_URL = 'http://www.google.com/get/noto/'
DESIGNER_URL = VENDOR_URL
LICENSE = ('This Font Software is licensed under the SIL Open Font License, '
           'Version 1.1. This Font Software is distributed on an "AS IS" '
           'BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either '
           'express or implied. See the SIL Open Font License for the '
           'specific language, permissions and limitations governing your '
           'use of this Font Software.')
LICENSE_URL = 'http://scripts.sil.org/OFL'
FSTYPE = 0  # Installable embedding

UPM = 2048
EMOJI_H_ADV = 2550
EMOJI_V_ADV = 2500
EMOJI_SIZE = 2400  # ASCENT + abs(DESCENT)
ABOVE_BASELINE = 0.7451  # ASCENT / EMOJI_H_ADV
ASCENT = 1900
DESCENT = -500
UNDERLINE_POSITION = -1244
UNDERLINE_THICKNESS = 131


SPACE_CHARSTRING = T2CharString(program=[EMOJI_H_ADV, 'endchar'])

RE_UNICODE = re.compile(r'^u[0-9a-f]{4,5}$', re.IGNORECASE)
RE_REVISION = re.compile(r'^[0-9]{1,3}\.[0-9]{3}$')
# The value of the viewBox attribute is a list of four numbers
# min-x, min-y, width and height, separated by whitespace and/or a comma
RE_VIEWBOX = re.compile(
    r"(<svg.+?)(\s*viewBox=[\"|\']([-\d,. ]+)[\"|\'])(.+?>)", re.DOTALL)

VALID_1STCHARS = tuple('_ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz')
VALID_CHARS = VALID_1STCHARS + tuple('.0123456789')

TAG_LAT_LETTR = ('e0061 e0062 e0063 e0064 e0065 e0066 e0067 e0068 e0069 e006a '
                 'e006b e006c e006d e006e e006f e0070 e0071 e0072 e0073 e0074 '
                 'e0075 e0076 e0077 e0078 e0079 e007a e007f').split()

log = logging.getLogger('make_bw_font')


def parse_viewbox_values(vb_str):
    """
    Input: viewbox's values string
    Return: list of integers or floats of viewbox's values
    """
    list_str = re.split(r'[\s,]', vb_str)
    assert len(list_str) == 4, 'viewBox must have 4 values'
    return [literal_eval(val) for val in list_str]


def get_svg_size(svg_file_path):
    """
    Takes a path to an SVG file and reads it.
    Checks for the existence of a 'viewBox' property in the 'svg' element.
    Confirms that the viewBox is square.
    Returns the viewBox dimension as an integer.

    The regex match will contain 4 groups:
        1. String from '<svg' up to the space before 'viewBox'
        2. The whole 'viewBox' property (e.g. ' viewBox="0 100 128 128"')
        3. The 'viewBox' values
        4. Remainder of the '<svg>' element
    """
    with io.open(svg_file_path, encoding='utf-8') as fp:
        svg_str = fp.read()

    vb = RE_VIEWBOX.search(svg_str)
    if not vb:
        log.error(f"'viewBox' property not found in {svg_file_path}")
        return

    min_x, min_y, width, height = parse_viewbox_values(vb.group(3))
    if not (min_x == min_y == 0):
        log.error("The origin of the 'viewBox' is not zero. "
                  f"min-x: {min_x}; min-y: {min_y}; {svg_file_path}")
        return

    if width != height:
        log.error("The 'viewBox' is not square. "
                  f"width: {width}; height: {height}; {svg_file_path}")
        return

    return width


def draw_notdef(pen):
    em_10th = EMOJI_H_ADV / 10
    v_shift = EMOJI_H_ADV * (ABOVE_BASELINE - 1)
    pen.moveTo((em_10th * 2, em_10th * 1 + v_shift))
    pen.lineTo((em_10th * 8, em_10th * 1 + v_shift))
    pen.lineTo((em_10th * 8, em_10th * 9 + v_shift))
    pen.lineTo((em_10th * 2, em_10th * 9 + v_shift))
    pen.closePath()
    pen.moveTo((em_10th * 3, em_10th * 2 + v_shift))
    pen.lineTo((em_10th * 3, em_10th * 8 + v_shift))
    pen.lineTo((em_10th * 7, em_10th * 8 + v_shift))
    pen.lineTo((em_10th * 7, em_10th * 2 + v_shift))
    pen.closePath()


def glyph_name_is_valid(gname, fpath):
    """
    Validates a string meant to be used as a glyph name, following the rules
    defined at https://adobe-type-tools.github.io/afdko/...
                       OpenTypeFeatureFileSpecification.html#2.f.i
    Returns True if the glyph name is valid and False otherwise.
    """
    if not gname:
        log.warning("Unable to get a glyph name from file '{}'.".format(fpath))
        return False
    elif gname[0] not in VALID_1STCHARS:
        log.warning("Glyph name made from file '{}' starts with an invalid "
                    "character '{}'.".format(fpath, gname[0]))
        return False
    elif not all([char in VALID_CHARS for char in tuple(gname)]):
        log.warning("Glyph name made from file '{}' contains one or more "
                    "invalid characters.".format(fpath))
        return False
    return True


def get_trimmed_glyph_name(gname, num):
    """
    Glyph names cannot have more than 31 characters.
    See https://docs.microsoft.com/en-us/typography/opentype/spec/...
               recom#39post39-table
    Trims an input string and appends a number to it.
    """
    suffix = '_{}'.format(num)
    return gname[:31 - len(suffix)] + suffix


def make_font(file_paths, out_dir, revision, gsub_path, gpos_path, uvs_lst):
    cmap, gorder, validated_fpaths = {}, deque(), []
    # build glyph order
    for fpath in file_paths:
        # derive glyph name from file name
        gname = os.path.splitext(os.path.basename(fpath))[0]  # trim extension
        # validate glyph name
        if not glyph_name_is_valid(gname, fpath):
            continue
        # skip any duplicates and 'space'
        if gname in gorder or gname == 'space':
            log.warning("Skipped file '{}'. The glyph name derived from it "
                        "is either a duplicate or 'space'.".format(fpath))
            continue
        # limit the length of glyph name to 31 chars
        if len(gname) > 31:
            num = 0
            trimmed_gname = get_trimmed_glyph_name(gname, num)
            while trimmed_gname in gorder:
                num += 1
                trimmed_gname = get_trimmed_glyph_name(trimmed_gname, num)
            gorder.append(trimmed_gname)
            log.warning("Glyph name '{}' was trimmed to 31 characters: "
                        "'{}'".format(gname, trimmed_gname))
        else:
            gorder.append(gname)
        validated_fpaths.append(fpath)

        # add to cmap
        if RE_UNICODE.match(gname):
            uni_int = int(gname[1:], 16)  # trim leading 'u'
            cmap[uni_int] = gname

    fb = FontBuilder(UPM, isTTF=False)
    fb.font['head'].fontRevision = float(revision)
    fb.font['head'].lowestRecPPEM = 12

    cs_dict = {}
    cs_cache = {}
    for i, svg_file_path in enumerate(validated_fpaths):
        svg_file_realpath = os.path.realpath(svg_file_path)

        if svg_file_realpath not in cs_cache:
            svg_size = get_svg_size(svg_file_realpath)
            if svg_size is None:
                cs_dict[gorder[i]] = SPACE_CHARSTRING
                continue

            pen = T2CharStringPen(EMOJI_H_ADV, None)
            svg = SVGPath(svg_file_realpath,
                          transform=(EMOJI_SIZE / svg_size, 0, 0,
                                     -EMOJI_SIZE / svg_size,
                                     (EMOJI_H_ADV * .5) - (EMOJI_SIZE * .5),
                                     EMOJI_H_ADV * ABOVE_BASELINE))
            svg.draw(pen)
            cs = pen.getCharString()
            cs_cache[svg_file_realpath] = cs
        else:
            cs = cs_cache.get(svg_file_realpath)

        cs_dict[gorder[i]] = cs

    # add '.notdef', 'space' and zero-width joiner
    pen = T2CharStringPen(EMOJI_H_ADV, None)
    draw_notdef(pen)
    gorder.extendleft(reversed(['.notdef', 'space', 'ZWJ']))
    cs_dict.update({'.notdef': pen.getCharString(),
                    'space': SPACE_CHARSTRING,
                    'ZWJ': SPACE_CHARSTRING,
                    })
    cmap.update({32: 'space',   # U+0020
                 160: 'space',  # U+00A0
                 8205: 'ZWJ',   # U+200D
                 })

    # add TAG LATIN LETTER glyphs and mappings
    for cdpt in TAG_LAT_LETTR:
        tag_gname = f'u{cdpt}'
        gorder.append(tag_gname)
        cs_dict[tag_gname] = SPACE_CHARSTRING
        cmap[int(cdpt, 16)] = tag_gname

    fb.setupGlyphOrder(list(gorder))  # parts of FontTools require a list
    fb.setupCharacterMap(cmap, uvs=uvs_lst)
    fb.setupCFF(PS_NAME, {'version': revision,
                          'Notice': TRADEMARK,
                          'Copyright': COPYRIGHT,
                          'FullName': FULL_NAME,
                          'FamilyName': FAMILY_NAME,
                          'Weight': STYLE_NAME}, cs_dict, {})

    glyphs_bearings = {}
    for gname, cs in cs_dict.items():
        gbbox = cs.calcBounds(None)
        if gbbox:
            xmin, ymin, _, ymax = gbbox
            if ymax > ASCENT:
                log.warning("Top of glyph '{}' may get clipped. "
                            "Glyph's ymax={}; Font's ascent={}".format(
                                gname, ymax, ASCENT))
            if ymin < DESCENT:
                log.warning("Bottom of glyph '{}' may get clipped. "
                            "Glyph's ymin={}; Font's descent={}".format(
                                gname, ymin, DESCENT))
            lsb = xmin
            tsb = EMOJI_V_ADV - ymax - EMOJI_H_ADV * (1 - ABOVE_BASELINE)
            glyphs_bearings[gname] = (lsb, tsb)
        else:
            glyphs_bearings[gname] = (0, 0)

    h_metrics = {}
    v_metrics = {}
    for gname in gorder:
        h_metrics[gname] = (EMOJI_H_ADV, glyphs_bearings[gname][0])
        v_metrics[gname] = (EMOJI_V_ADV, glyphs_bearings[gname][1])
    fb.setupHorizontalMetrics(h_metrics)
    fb.setupVerticalMetrics(v_metrics)

    fb.setupHorizontalHeader(ascent=ASCENT, descent=DESCENT)

    v_ascent = EMOJI_H_ADV // 2
    v_descent = EMOJI_H_ADV - v_ascent
    fb.setupVerticalHeader(
        ascent=v_ascent, descent=-v_descent, caretSlopeRun=1)

    VERSION_STRING = 'Version {};{}'.format(revision, VENDOR)
    UNIQUE_ID = '{};{};{}'.format(revision, VENDOR, PS_NAME)
    name_strings = dict(
        copyright=COPYRIGHT,             # ID 0
        familyName=FAMILY_NAME,          # ID 1
        styleName=STYLE_NAME,            # ID 2
        uniqueFontIdentifier=UNIQUE_ID,  # ID 3
        fullName=FULL_NAME,              # ID 4
        version=VERSION_STRING,          # ID 5
        psName=PS_NAME,                  # ID 6
        trademark=TRADEMARK,             # ID 7
        manufacturer=MANUFACTURER,       # ID 8
        designer=DESIGNER,               # ID 9
        vendorURL=VENDOR_URL,            # ID 11
        designerURL=DESIGNER_URL,        # ID 12
        licenseDescription=LICENSE,      # ID 13
        licenseInfoURL=LICENSE_URL,      # ID 14
    )
    fb.setupNameTable(name_strings, mac=False)

    fb.setupOS2(fsType=FSTYPE, achVendID=VENDOR, fsSelection=0x0040,  # REGULAR
                usWinAscent=ASCENT, usWinDescent=-DESCENT,
                sTypoAscender=ASCENT, sTypoDescender=DESCENT,
                sCapHeight=ASCENT, ulCodePageRange1=(1 << 1))  # set 1st CP bit

    if gsub_path:
        addOpenTypeFeatures(fb.font, gsub_path, tables=['GSUB'])

    if gpos_path:
        addOpenTypeFeatures(fb.font, gpos_path, tables=['GPOS'])

    fb.setupPost(isFixedPitch=1,
                 underlinePosition=UNDERLINE_POSITION,
                 underlineThickness=UNDERLINE_THICKNESS)

    fb.setupDummyDSIG()

    fb.save(os.path.join(out_dir, '{}.otf'.format(PS_NAME)))


def parse_uvs_file(file_path):
    """
    Parses an Unicode Variation Sequences text file.
    Returns a list of tuples in the form
    (unicodeValue, variationSelector, glyphName).
    'unicodeValue' and 'variationSelector' are integer code points.
    'glyphName' may be None, to indicate this is the default variation.
    """
    with io.open(file_path, encoding='utf-8') as fp:
        lines = fp.read().splitlines()

    uvs_list = []
    for i, line in enumerate(lines, 1):
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        uni_str, gname = line.split(';')
        uni_lst = uni_str.strip().split()
        if not isinstance(uni_lst, list) and len(uni_lst) != 2:
            log.error('Line #{} is not correctly formatted.'.format(i))
            continue
        try:
            uni_int = [int(cdpt, 16) for cdpt in uni_lst]
        except ValueError:
            log.error('Line #{} has an invalid code point.'.format(i))
            continue
        gname = gname.strip()
        if gname == 'None':
            gname = None
        uni_int.append(gname)
        uvs_item = tuple(uni_int)
        if uvs_item in uvs_list:
            log.warning('Line #{} is a duplicate UVS.'.format(i))
            continue
        uvs_list.append(uvs_item)
    if not uvs_list:
        log.warning('No Unicode Variation Sequences were found.')
        return None
    return uvs_list


def validate_dir_path(path_str):
    valid_path = os.path.abspath(os.path.realpath(path_str))
    if not os.path.isdir(valid_path):
        raise argparse.ArgumentTypeError(
            "{} is not a valid directory path.".format(path_str))
    return normalize_path(path_str)


def validate_file_path(path_str):
    valid_path = os.path.abspath(os.path.realpath(path_str))
    if not os.path.isfile(valid_path):
        raise argparse.ArgumentTypeError(
            "{} is not a valid file path.".format(path_str))
    return normalize_path(path_str)


def normalize_path(path_str):
    return os.path.normpath(path_str)


def validate_revision_number(rev_str):
    if not RE_REVISION.match(rev_str):
        raise argparse.ArgumentTypeError(
            "The revision number must follow this format: 123.456")
    return rev_str


def main(args=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        '-v',
        '--verbose',
        help='verbose mode. Use -vv for debug mode',
        action='count',
        default=0
    )
    parser.add_argument(
        'in_dirs',
        help='one or more input directories containing SVG files',
        metavar='DIR',
        nargs='+',
        type=validate_dir_path,
    )
    parser.add_argument(
        '-o',
        '--out-dir',
        help='directory to save the font in. Defaults to 1st input directory.',
        metavar='DIR',
        type=normalize_path,
    )
    parser.add_argument(
        '-r',
        '--revision',
        help="the font's revision number. Defaults to %(default)s",
        type=validate_revision_number,
        default='0.001',
    )
    parser.add_argument(
        '--gsub',
        help='path to GSUB features file',
        type=validate_file_path,
    )
    parser.add_argument(
        '--gpos',
        help='path to GPOS features file',
        type=validate_file_path,
    )
    parser.add_argument(
        '--uvs',
        help='path to Unicode Variation Sequences file',
        type=validate_file_path,
    )
    opts = parser.parse_args(args)

    if not opts.verbose:
        level = "WARNING"
    elif opts.verbose == 1:
        level = "INFO"
    else:
        level = "DEBUG"
    logging.basicConfig(level=level)

    file_paths = []
    for in_dir in opts.in_dirs:
        fpaths = sorted(glob.iglob(os.path.join(in_dir, '*.[sS][vV][gG]')))
        file_paths.extend(fpaths)
        log.info(f"Found {len(fpaths)} SVG files in '{in_dir}'.")

    if not len(file_paths):
        log.error('Failed to match any SVG files.')
        return 1

    uvs = None
    if opts.uvs:
        uvs = parse_uvs_file(opts.uvs)

    if opts.out_dir:
        out_path = os.path.abspath(os.path.realpath(opts.out_dir))
        # create directory if it doesn't exist
        if not os.path.exists(out_path):
            os.makedirs(out_path)
        # the path exists but it's NOT a directory
        elif not os.path.isdir(out_path):
            log.error("'{}' is not a directory.".format(opts.out_dir))
            return 1
        out_dir = opts.out_dir
    else:
        out_dir = opts.in_dirs[0]

    make_font(file_paths, out_dir, opts.revision, opts.gsub, opts.gpos, uvs)


if __name__ == "__main__":
    sys.exit(main())
